python/whylogs/api/writer/whylabs.py

```diff
@@ -82,17 +145,105 @@ class WhyLabsWriter(Writer):
+        self._endpoint_hostname = os.environ.get("WHYLABS_ENDPOINT_HOSTNAME_OVERRIDE") or None
         self.whylabs_api_endpoint = os.environ.get("WHYLABS_API_ENDPOINT") or "https://api.whylabsapp.com"
         self._reference_profile_name = os.environ.get("WHYLABS_REFERENCE_PROFILE_NAME")
-        self._api_log_client: Optional[ApiClient] = None
-        self._ssl_ca_cert: Optional[str] = None
-        self._configuration: Optional[Configuration] = None
+        self._ssl_ca_cert = ssl_ca_cert
+        self._api_config: Optional[Configuration] = None
+        if api_key:
+            self._key_refresher = StaticKeyRefresher(api_key)
+        else:
+            self._key_refresher = _ENV_KEY_REFRESHER
+        if api_client:
+            self._api_client = api_client
+        else:
+            self._api_client = None
+            self._refresh_client()
+
+    @property
+    def key_id(self) -> str:
+        return self._key_refresher.key_id
+
+    def _refresh_client(self) -> None:
+        """
+        Refresh the API client by comparing various configs. We try to
+        re-use the client as much as we can since using a new client
+        every time can be expensive.
+
+        """
+        cache_key = ""
+
+        if self._api_client:
+            config = copy.deepcopy(self._api_client.configuration)
+        else:
+            config = Configuration()
+        # Set an empty api key. The key refresher will refresh it
+        config.api_key = {"ApiKeyAuth": ""}
+        config.refresh_api_key_hook = self._key_refresher
+        cache_key += str(hash(self._key_refresher))
+
+        config.discard_unknown_keys = True
+        # Disable client side validation and trust the server
+        config.client_side_validation = False
+
+        cache_key += str(hash(config))
+        if self._ssl_ca_cert:
+            config.ssl_ca_cert = self._ssl_ca_cert
+            cache_key += str(hash(self._ssl_ca_cert))
+        config.host = self.whylabs_api_endpoint
+        cache_key += str(hash(self.whylabs_api_endpoint))
+        if self._endpoint_hostname:
+            cache_key += str(hash(self._endpoint_hostname))
+
+        existing_client = _API_CLIENT_CACHE.get(cache_key)
+        if existing_client:
+            logger.debug(f"Found existing client under cache key: {cache_key}")
+            self._api_client = existing_client
+            return
+
+        client = whylabs_client.ApiClient(config)
+        client.user_agent = f"whylogs/python/{_version}"
+        if self._endpoint_hostname:
+            self._update_hostname_config(self._endpoint_hostname)
+
+        self._api_client = client
+        _API_CLIENT_CACHE[cache_key] = client
+        logger.debug(f"Created and updated new client for cache key: {cache_key}")
+
+    def _update_hostname_config(self, endpoint_hostname_override: str) -> None:
+        """
+        This method overrides the pool manager's new connection method to add the hostname
+
+        """
+        import urllib3
```

Comment:
Do you want to add urllib3 as a requirement in the whylabs extra in pyproject.toml file (next to requests and whylabs-client)?